File: v1/examples/swift/02-channels.swift
/**
 * Example 02: Channel Management
 *
 * Demonstrates:
 *   - List channels (basic info, static source, capabilities)
 *   - Channel status (connection state, recording status)
 *   - Channel detailed info
 */

import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

@main
struct App {
  static func main() async throws {
    let client = NvrClient()
    guard !client.apiKey.isEmpty else {
        print("NVR_API_KEY environment variable is required")
        exit(1)
    }
    client.setApiKey(client.apiKey)

    // -------------------------------------------------
    // 1. List Channels (basic info)
    // -------------------------------------------------
    print("=== Channel List ===")
    let channelRes = try await client.get("/api/v1/channel")

    if channelRes.status == 200, let channels = channelRes.body as? [[String: Any]] {
        for ch in channels {
            let chid = ch["chid"] ?? ""
            let title = ch["title"] as? String ?? ""
            let displayName = ch["displayName"] as? String ?? ""
            print("  CH\(chid): \(title) (\(displayName))")
        }
        print("Total: \(channels.count) channels\n")
    }

    // -------------------------------------------------
    // 2. List Channels with static source URLs
    // -------------------------------------------------
    print("=== Channel List with Sources ===")
    let srcRes = try await client.get("/api/v1/channel?staticSrc")

    if srcRes.status == 200, let channels = srcRes.body as? [[String: Any]] {
        for ch in channels {
            let chid = ch["chid"] ?? ""
            let title = ch["title"] as? String ?? ""
            print("  CH\(chid): \(title)")
            if let src = ch["src"] {
                print("    Source: \(client.toJsonString(src))")
            }
        }
        print("")
    }

    // -------------------------------------------------
    // 3. List Channels with capabilities
    // -------------------------------------------------
    print("=== Channel Capabilities ===")
    let capsRes = try await client.get("/api/v1/channel?caps")

    if capsRes.status == 200, let channels = capsRes.body as? [[String: Any]] {
        for ch in channels {
            let chid = ch["chid"] ?? ""
            let title = ch["title"] as? String ?? ""
            var features: [String] = []
            if let caps = ch["caps"] as? [String: Any] {
                if caps["pantilt"] as? Bool == true { features.append("Pan/Tilt") }
                if caps["zoom"] as? Bool == true { features.append("Zoom") }
                if caps["relay"] as? Bool == true { features.append("Relay") }
            }
            let featureStr = features.isEmpty ? "No PTZ" : features.joined(separator: ", ")
            print("  CH\(chid): \(title) [\(featureStr)]")
        }
        print("")
    }

    // -------------------------------------------------
    // 4. Channel Status
    // -------------------------------------------------
    print("=== Channel Status ===")
    let statusRes = try await client.get("/api/v1/channel/status?recordingStatus")

    if statusRes.status == 200, let channels = statusRes.body as? [[String: Any]] {
        // Status codes: 0=Connected, -1=Disconnected, -2=Connecting, -3=Auth Failed
        let statusMap: [Int: String] = [0: "Connected", -1: "Disconnected", -2: "Connecting", -3: "Auth Failed"]

        for ch in channels {
            let chid = ch["chid"] ?? ""
            var statusText = "Unknown"
            if let status = ch["status"] as? [String: Any], let code = status["code"] as? Int {
                statusText = statusMap[code] ?? "Unknown(\(code))"
            }
            let recording = ch["recording"] as? Bool ?? false
            let recText = recording ? " [REC]" : ""
            print("  CH\(chid): \(statusText)\(recText)")
        }
        print("")
    }

    // -------------------------------------------------
    // 5. Specific Channel Info
    // -------------------------------------------------
    print("=== Channel 1 Detailed Info ===")
    let infoRes = try await client.get("/api/v1/channel/1/info?caps")
    print("  Status: \(infoRes.status)")
    if infoRes.status == 200, let body = infoRes.body {
        print("  Data: \(client.toJsonString(body, pretty: true))")
    }
  }
}

File: v1/examples/javascript/02-channels.js
/**
 * Example 02: Channel Management
 *
 * Demonstrates:
 *   - List channels (basic info, static source, capabilities)
 *   - Channel status (connection state, recording status)
 *   - Channel detailed info
 */

const { NVR_API_KEY } = require('./config');
const { setApiKey, get } = require('./http');

async function main() {
  if (!NVR_API_KEY) {
    console.error('NVR_API_KEY environment variable is required');
    process.exit(1);
  }
  setApiKey(NVR_API_KEY);

  // ─────────────────────────────────────────────────
  // 1. List Channels (basic info)
  // ─────────────────────────────────────────────────
  console.log('=== Channel List ===');
  const channelRes = await get('/api/v1/channel');

  if (channelRes.status === 200 && Array.isArray(channelRes.body)) {
    for (const ch of channelRes.body) {
      // Response fields: chid, title, displayName
      console.log(`  CH${ch.chid}: ${ch.title} (${ch.displayName})`);
    }
    console.log(`Total: ${channelRes.body.length} channels\n`);
  }

  // ─────────────────────────────────────────────────
  // 2. List Channels with static source URLs
  // ─────────────────────────────────────────────────
  console.log('=== Channel List with Sources ===');
  const srcRes = await get('/api/v1/channel?staticSrc');

  if (srcRes.status === 200 && Array.isArray(srcRes.body)) {
    for (const ch of srcRes.body) {
      console.log(`  CH${ch.chid}: ${ch.title}`);
      if (ch.src) {
        console.log(`    Source: ${JSON.stringify(ch.src)}`);
      }
    }
    console.log('');
  }

  // ─────────────────────────────────────────────────
  // 3. List Channels with capabilities
  // ─────────────────────────────────────────────────
  console.log('=== Channel Capabilities ===');
  const capsRes = await get('/api/v1/channel?caps');

  if (capsRes.status === 200 && Array.isArray(capsRes.body)) {
    for (const ch of capsRes.body) {
      const features = [];
      if (ch.caps) {
        if (ch.caps.pantilt) features.push('Pan/Tilt');
        if (ch.caps.zoom) features.push('Zoom');
        if (ch.caps.relay) features.push('Relay');
      }
      console.log(`  CH${ch.chid}: ${ch.title} [${features.join(', ') || 'No PTZ'}]`);
    }
    console.log('');
  }

  // ─────────────────────────────────────────────────
  // 4. Channel Status
  // ─────────────────────────────────────────────────
  console.log('=== Channel Status ===');
  const statusRes = await get('/api/v1/channel/status?recordingStatus');

  if (statusRes.status === 200 && Array.isArray(statusRes.body)) {
    // Status codes: 0=Connected, -1=Disconnected, -2=Connecting, -3=Auth Failed
    const statusMap = { 0: 'Connected', '-1': 'Disconnected', '-2': 'Connecting', '-3': 'Auth Failed' };

    for (const ch of statusRes.body) {
      const statusText = statusMap[ch.status?.code] || `Unknown(${ch.status?.code})`;
      const recText = ch.recording ? ' [REC]' : '';
      console.log(`  CH${ch.chid}: ${statusText}${recText}`);
    }
    console.log('');
  }

  // ─────────────────────────────────────────────────
  // 5. Specific Channel Info
  // ─────────────────────────────────────────────────
  console.log('=== Channel 1 Detailed Info ===');
  const infoRes = await get('/api/v1/channel/1/info?caps');
  console.log(`  Status: ${infoRes.status}`);
  if (infoRes.status === 200) {
    console.log(`  Data: ${JSON.stringify(infoRes.body, null, 2)}`);
  }

}

main().catch(console.error);
